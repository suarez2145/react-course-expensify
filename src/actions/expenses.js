import uuid from 'uuid';
import database from '../firebase/firebase';




// ADD_EXPENSE action function

// this action function is now getting its data from the stratAddExpense function below 
export const AddExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})


export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData;
        const expense = {description, note, amount, createdAt};
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(AddExpense({
                id: ref.key,
                ...expense
            }))
        });
    }
}

// REMOVE_EXPENSE action function

export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

// EDIT_EXPENSE action function

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// this action generator is the one that is responsible for retrieving the expenses data from firebase so that our 
// data is displayed and saved to the screen 

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expenses))
        })
    }
}