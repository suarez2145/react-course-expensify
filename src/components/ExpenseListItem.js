import React from 'react';
import { Link } from 'react-router-dom';



// create our function that takes in the destructured object (our expenses object)
// here as arguments we pass the variables we are going to need (description,amount,createdAt)
// now that we also imported { connect } that gives us access to the store " dispatch" function
// we include it in the object we pass to ExpenseListItem function, we also pass in "id" so we can use the removeExpense function in our dispatch call 
const ExpenseListItem = ({id, description, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>{amount} - {createdAt}</p>
    </div>
);

// we export so we can import on another file and use there.

// 
export default ExpenseListItem