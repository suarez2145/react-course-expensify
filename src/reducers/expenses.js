



// Expenses Reducer below

// creating a variable to hold our DEFAULTs for demoState (because passing them into expenses reducers as inline arguments might get to long )
const expensesReducerDefaultState = [];


// Expenses Reducer below
const expensesReducer = (state = expensesReducerDefaultState , action) => {
    switch (action.type) {

        case 'ADD_EXPENSE':
            return [...state, action.expense]
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => {
                return id !== action.id;
            })
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id){
                    return {
                    ...expense,
                    ...action.updates
                    }

                } else {
                    return expense;
                }
            })
        default: 
        return state;
    }
}

export default expensesReducer;