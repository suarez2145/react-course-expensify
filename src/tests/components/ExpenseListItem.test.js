import React from 'react';
import { shallow } from 'enzyme';
import  ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('should render ExpenseListItem with fixture item data', () => {
    // below the ExpenseListItem component expected an object with (id, description, amount, createAt) to be passed in 
    // so we use the spread operator to access these variables and vaules from the first expenses item
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>)
    expect(wrapper).toMatchSnapshot();
})