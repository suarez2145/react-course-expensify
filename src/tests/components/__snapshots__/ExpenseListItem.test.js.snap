// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with fixture item data 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    $1.95
    -
    December 31st, 1969
  </p>
</div>
`;
